contrib/nydusify/pkg/converter/reverse_converter.go

```diff
@@ -0,0 +1,597 @@
+	if err != nil {
+		return errors.Wrap(err, "push oci image")
+	}
+
+	logrus.Infof("Successfully converted Nydus image %s to OCI image %s", opt.Source, opt.Target)
+	return nil
+}
+
+// reverseHosts creates host configuration for reverse conversion
+func reverseHosts(opt ReverseOpt) func(string) (func(string) (string, string, error), bool, error) {
+	maps := map[string]bool{
+		opt.Source: opt.SourceInsecure,
+		opt.Target: opt.TargetInsecure,
+	}
+	return func(ref string) (func(string) (string, string, error), bool, error) {
+		return func(string) (string, string, error) { return "", "", nil }, maps[ref], nil
+	}
+}
+
+// pullNydusImage pulls Nydus image and extracts layer information
+func pullNydusImage(ctx context.Context, opt ReverseOpt, pvd *provider.Provider, tmpDir string) ([]ocispec.Descriptor, *ocispec.Manifest, error) {
+	logrus.Infof("Pulling Nydus image: %s", opt.Source)
+
+	// Create remote for source
+	remoter, err := pkgPvd.DefaultRemote(opt.Source, opt.SourceInsecure)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "create source remote")
+	}
+
+	if opt.WithPlainHTTP {
+		remoter.WithHTTP()
+	}
+
+	// Resolve manifest
+	manifestDesc, err := remoter.Resolve(ctx)
+	if utils.RetryWithHTTP(err) {
+		remoter.MaybeWithHTTP(err)
+		manifestDesc, err = remoter.Resolve(ctx)
+	}
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "resolve nydus manifest")
+	}
+
+	// Pull manifest
+	manifestReader, err := remoter.Pull(ctx, *manifestDesc, true)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "pull nydus manifest")
+	}
+	defer manifestReader.Close()
+
+	manifestBytes, err := io.ReadAll(manifestReader)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "read manifest bytes")
+	}
+
+	var manifest ocispec.Manifest
+	if err := json.Unmarshal(manifestBytes, &manifest); err != nil {
+		return nil, nil, errors.Wrap(err, "unmarshal manifest")
+	}
+
+	// Pull all layers
+	var layers []ocispec.Descriptor
+	for _, layer := range manifest.Layers {
+		layerPath := filepath.Join(tmpDir, layer.Digest.Hex())
+		layerFile, err := os.Create(layerPath)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "create layer file %s", layerPath)
+		}
+
+		layerReader, err := remoter.Pull(ctx, layer, true)
+		if err != nil {
+			layerFile.Close()
+			return nil, nil, errors.Wrapf(err, "pull layer %s", layer.Digest)
+		}
+
+		_, err = io.Copy(layerFile, layerReader)
+		layerReader.Close()
+		layerFile.Close()
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "copy layer %s", layer.Digest)
+		}
+
+		layers = append(layers, layer)
+	}
+
+	return layers, &manifest, nil
+}
+
+// unpackNydusLayers unpacks Nydus layers using nydus-image unpack command
+func unpackNydusLayers(ctx context.Context, opt ReverseOpt, tmpDir string, nydusLayers []ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	logrus.Info("Unpacking Nydus layers to OCI format")
+
+	var ociLayers []ocispec.Descriptor
+
+	for i, layer := range nydusLayers {
```

Comment:
It may be slow when processing large images, so consider using goroutine to process layers.

Reply:
Let's refactor the workflow codes later. @J-jxr